### PR TITLE
Optimize WarpAffine(linear) using AVX2

### PR DESCRIPTION
**Pull request by woodychow**
Pretty straight forward. I just replaced the SSE instructions with the AVX2 instructions. I added this line

```cpp
fx_ = _mm256_permute4x64_epi64(fx_, (3 << 6) + (1 << 4) + (2 << 2) + 0);
```
to put fx_in the right order.

### CODE REVIEW

File: modules/imgproc/src/imgwarp.cpp [hunk 1]
```diff
@@ -5603,7 +5605,41 @@ class WarpAffineInvoker :
...
+                                                         _mm256_srai_epi32(ty1, INTER_BITS));
+                                fx_ = _mm256_adds_epi16(fx_, _mm256_slli_epi16(fy_, INTER_BITS));
+                                fx_ = _mm256_permute4x64_epi64(fx_, (3 << 6) + (1 << 4) + (2 << 2) + 0);
+
+                                _mm256_storeu_si256((__m256i*)(xy + x1*2), _mm256_unpacklo_epi16(tx0, ty0));
+                                _mm256_storeu_si256((__m256i*)(xy + x1*2 + 16), _mm256_unpackhi_epi16(tx0, ty0));
+                                _mm256_storeu_si256((__m256i*)(alpha + x1), fx_);
+                            }
+                        }
+                    #elif CV_SSE2
```
Comment (alalek) [inline modules/imgproc/src/imgwarp.cpp]: Probably we should not drop SSE2 optimization if we want to add AVX2 optimization. Also we should use vzeroupper if we [mix AVX/SSE](http://stackoverflow.com/questions/41303780/why-is-this-sse-code-6-times-slower-without-vzeroupper-on-skylake) instructions.
```.diff
         }
+        _mm256_zeroupper();
     }
-#elif CV_SSE2
+#endif
+#if CV_SSE2
```

Comment (woodychow) [inline modules/imgproc/src/imgwarp.cpp]: OK I changed the code. Thanks a lot for the link. I did not know that.

**Comment by woodychow** [comment]
The server ran out of space. How can the build be re-triggered?

```
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1097, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/lib/python2.7/dist-packages/twisted/spread/pb.py", line 470, in throwExceptionIntoGenerator
    return g.throw(RemoteError(self.type, self.value, self.traceback))
  File "/app/buildbot/master/buildbot/process/buildstep.py", line 583, in runCommand
    res = yield command.run(self, self.remote)
twisted.spread.pb.RemoteError: [Errno 28] No space left on device
```

**Comment by vpisarev** [comment]
thank you! can you, please, refactor your code to support runtime dispatching? See https://github.com/opencv/opencv/pull/8399 for details

**Comment by vpisarev** [comment]
should we close this one, since @alalek reworked it?

**Comment by woodychow** [comment]
I thought he only reworked #8392. We can close #8392.

**Comment by alalek** [comment]
No need to close anything (including #8392).

**Comment by vpisarev** [comment]
let's put it in while there are no merge conflicts! :+1:

**Pull request merged**